File: LeetCode/0066. Plus One/Solution.py
from functools import reduce
class Solution:
    def plusOne(self, digits):
        """
        109 / 109 test cases passed.
        Status: Accepted
        Runtime: 40 ms
        ----------------
        :type digits: List[int]
        :rtype: List[int]
        """
        num = reduce(lambda acc, x: acc * 10 + x, digits)
        num += 1
        return list(map(lambda x: int(x), str(num)))

class Solution2:
    """
    111 / 111 test cases passed.
    Status: Accepted
    Runtime: 32 ms
    ----------------
    :type digits: List[int]
    :rtype: List[int]
    """
    def plusOne(self, digits: List[int]) -> List[int]:
        n = len(digits)
        for i in range(n - 1, -1, -1):
            if digits[i] == 9:
                digits[i] = 0
            else:
                digits[i] += 1
                return digits
        return [1] + [0] * n

File: LeetCode/0066. Plus One/Solution.cpp
// 109 / 109 test cases passed.
// Status: Accepted
// Runtime: 4 ms

class Solution {
public:
    vector<int> plusOne(vector<int>& digits) {
        for (int i = digits.size() - 1; i >= 0; -- i) {
            if (digits[i] < 9) {
                digits[i] ++;
                return digits;
            } else {
                digits[i] = 0;
            }
        } 
        if (digits[0] == 0) digits.insert(digits.begin(), 1);
        return digits;
    }
};

// 111 / 111 test cases passed.
// Status: Accepted
// Runtime: 4 ms
class Solution2 {
public:
    vector<int> plusOne(vector<int>& digits) {
        int n = digits.size();
        for (int i = n - 1; i >= 0; i--) {
            if (digits[i] == 9) digits[i] = 0;
            else {
                digits[i] += 1;
                return digits;
            }
        }  
        vector<int> ans(n + 1);
        ans[0] = 1;
        return ans;
    }
};
